README.md
# generate-api-key  

[![GitHub package.json version (branch)][version-image]][npm-url]
[![unit tests][tests-image]][tests-url]
[![Build Status][travis-image]][travis-url]
[![coverage][coverage-image]][coverage-url]
[![Coverage Status][coveralls-image]][coveralls-url]
[![CodeFactor][codefactor-image]][codefactor-url]  


`generate-api-key` is a library for generating random API (Application Programming Interface) 
keys or access tokens. By using this library, a Node.js backend service can generate API keys 
or access tokens, then issue them to users and/or other services that require access to the 
capabilities and resources provided by the API service.  


## Table of contents

- [Installation](#installation)
- [Usage](#usage)
  - [Generation Methods](#generation-methods)
- [Options](#options)
  - [`string` Method](#string-method)
  - [`bytes` Method](#bytes-method)
  - [`base32` Method](#base32-method)
  - [`base62` Method](#base62-method)
  - [`uuidv4` Method](#uuidv4-method)
  - [`uuidv5` Method](#uuidv5-method)
- [Security](#security)
- [Change Log](#change-log)
- [License](#license)  
<br />

## Installation

Using NPM:

```bash
$ npm install generate-api-key
```

Using Yarn:

```bash
$ yarn add generate-api-key
```
<br />

## Usage

The `generate-api-key` library can generate API key/access tokens by utilizing several 
generation methods, such as `string`, `bytes`, `base32`, `base62`, `uuidv4`, and 
`uuidv5`. The `string` method is used by default.  
<br />

Importing:  

```javascript
// CommonJS Import
const { generateApiKey } = require('generate-api-key');
// OR
const generateApiKey = require('generate-api-key').default;

// ES6 Import
import { generateApiKey } from 'generate-api-key';
// OR
import generateApiKey from 'generate-api-key';
```

Example:

```javascript
import generateApiKey from 'generate-api-key';

// Generate the API key.
generateApiKey(); // ⇨ 'q_EaTiX+xbBXLyO05.+zDXjI+Qi_X0v'
```
<br />

### Generation Methods

| Method   |  Description                                                           |
| -------- | ---------------------------------------------------------------------- | 
| `string` | Creates an API key/access token using random string generation         |
| `bytes`  | Creates an API key/access token using random bytes                     |
| `base32` | Creates an API key/access token using a random UUID and converting it<br />into a [Douglas Crockford Base32](https://en.wikipedia.org/wiki/Base32#Crockford's_Base32) encoded string  |
| `base62` | Creates an API key using Base62 encoding                               |
| `uuidv4` | Creates an API key/access token using random UUID Version 4 generation |
| `uuidv5` | Creates an API key/access token using random UUID Version 5 generation |
<br />

## Options

### `string` Method  

Creates an API key/access token using random string generation.  

| Name     | Default Value |  Description                                                      |
| ---------| ------------- | ----------------------------------------------------------------- | 
| `method` | `string`      | To use the `string` generation method                             |
| `min`    | `16`          | The minimum length of the API key (ignored if `length` is given)  |
| `max`    | `32`          | The maximum length of the API key (ignored if `length` is given)  |
| `length` | `undefined`   | The length of the API key                                         |
| `pool`   | `abcdefghijklmnopqrstuvwxyz`<br />`ABCDEFGHIJKLMNOPQRSTUVWXYZ`<br />`0123456789-._~+/` | The characters used for the API key generation | 
| `prefix` | `undefined`   | A string prefix for the API key, followed by a period (`.`)       |  
| `batch`  | `undefined`   | The number of API keys to generate                                |
<br />

Examples:  

```javascript
import generateApiKey from 'generate-api-key';

// Generate the API key. The 'string' method is used by default.
generateApiKey(); // ⇨ 'q_EaTiX+xbBXLyO05.+zDXjI+Qi_X0v'

// Provide the generation method.
generateApiKey({ method: 'string' }); // ⇨ 'Zt1HbMcLKxk6~nnW'

// Create an API key with a certain length.
generateApiKey({ method: 'string', length: 8 }); // ⇨ 'TNJ6-Lo4'

// Create an API key with a length between a certain range.
generateApiKey({ method: 'string', min: 10, max: 20 }); // ⇨ 'ENwiOFdP8cWj'

// Create an API key with a certain pool of characters.
generateApiKey({
  method: 'string',
  pool: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
}); // ⇨ 'QFLSGIDLOUAELQZTQXMHQNJ'

// Create an API key with a prefix.
generateApiKey({ method: 'string', prefix: 'test_app' }); // ⇨ 'test_app.aTd34Rli0nir70/8'

// Create a batch (certain amount) of API keys.
generateApiKey({ method: 'string', batch: 5 }); // ⇨ 
  // [
  //   'w05KkI9AWhKxzvPFtXotUva-',
  //   'YFL0ICl4PtLD8Y/oQ20iyAE',
  //   'vJFbfeP_cpMYsH9l5BVHY23Ss',
  //   '29~LIlSjDYFr5OrhU3f',
  //   'UQc8Tp1d9elWAh7KDIMkjz2moFs'
  // ]
```
<br />

### `bytes` Method  

Creates an API key/access token using random bytes.  

| Name     | Default Value |  Description                                                      |
| ---------| ------------- | ----------------------------------------------------------------- | 
| `method` | `bytes`       | To use the `bytes` generation method                              |
| `min`    | `16`          | The minimum length of the API key (ignored if `length` is given)  |
| `max`    | `32`          | The maximum length of the API key (ignored if `length` is given)  |
| `length` | `undefined`   | The length of the API key                                         |
| `prefix` | `undefined`   | A string prefix for the API key, followed by a period (`.`)       |  
| `batch`  | `undefined`   | The number of API keys to generate                                |
<br />

Examples:  

```javascript
import generateApiKey from 'generate-api-key';

// Provide the generation method.
generateApiKey({ method: 'bytes' }); // ⇨ '6f31bfc3717d63e7bd21'

// Create an API key with a certain length.
generateApiKey({ method: 'bytes', length: 12 }); // ⇨ '47a8dcbc79f6'

// Create an API key with a length between a certain range.
generateApiKey({ method: 'bytes', min: 12, max: 25 }); // ⇨ 'fae27c801b5092bc'

// Create an API key with a prefix.
generateApiKey({ method: 'bytes', prefix: 'test_app' }); // ⇨ 'test_app.8daaa6b26c79030db1a1448261'

// Create a batch (certain amount) of API keys.
generateApiKey({ method: 'bytes', batch: 5 }); // ⇨ 
  // [
  //   '0d5a87f007aae092a6',
  //   '96a62b4438d82645506b',
  //   'abd4e4199311fb1e2a818a4a',
  //   'ddbb04b2375ba050cb506e89df',
  //   '2ee3db86329865d8'
  // ]
```
<br />

### `base32` Method  

Creates an API key/access token using a random UUID and converting it into a [Douglas Crockford Base32](https://en.wikipedia.org/wiki/Base32#Crockford's_Base32) encoded string.  

| Name     | Default Value |  Description                                                      |
| ---------| ------------- | ----------------------------------------------------------------- | 
| `method` | `base32`      | To use the `base32` generation method                             |
| `dashes` | `true`        | Add dashes (`-`) to the API key or not                            |
| `prefix` | `undefined`   | A string prefix for the API key, followed by a period (`.`)       |  
| `batch`  | `undefined`   | The number of API keys to generate                                |
<br />

Examples:  

```javascript
import generateApiKey from 'generate-api-key';

// Provide the generation method.
generateApiKey({ method: 'base32' }); // ⇨ '2NOLH5I-43EEK7A-R6YRK3I-BRCIQNQ'

// Create an API key without the dashes.
generateApiKey({ method: 'base32', dashes: false }); // ⇨ 'MPZ6G2QTAXUGBYTCLKEGYZG6UXAY'

// Create an API key with a prefix.
generateApiKey({ method: 'base32', prefix: 'test_app' }); // ⇨ 'test_app.PMKC6DQ-2LZECPY-RVRF6YI-INGIEMQ'

// Create a batch (certain amount) of API keys.
generateApiKey({ method: 'base32', batch: 5 }); // ⇨ 
  // [
  //   'EGIKGTI-PTLUW7I-QBLUBNA-FH4P2EI',
  //   'MOHE7XQ-DRKEQ4I-RO6JVDQ-GIJRLKA',
  //   'UDVJBFI-JFTUWYA-XE4YCBA-FUYXQBY',
  //   '22EFRQI-2AWULDI-QE2PUCY-MXG36RI',
  //   '42CXULQ-SDPU6ZA-RQ6QPBQ-4BMZCOA'
  // ]
```
<br />

### `base62` Method  

Creates an API key using Base62 encoding.  

| Name     | Default Value |  Description                                                      |
| ---------| ------------- | ----------------------------------------------------------------- | 
| `method` | `base62`      | To use the `base62` generation method                             |
| `prefix` | `undefined`   | A string prefix for the API key, followed by a period (`.`)       |  
| `batch`  | `undefined`   | The number of API keys to generate                                |
<br />

Examples:  

```javascript
import generateApiKey from 'generate-api-key';

// Provide the generation method.
generateApiKey({ method: 'base62' }); // ⇨ '2AEmXhHtNJkIAqL1S3So6G'

// Create an API key with a prefix.
generateApiKey({ method: 'base62', prefix: 'test_app' }); // ⇨ 'test_app.1aHVuwNwzITpzJWl40OPvx'

// Create a batch (certain amount) of API keys.
generateApiKey({ method: 'base62', batch: 5 }); // ⇨ 
  // [
  //   '1kCFJsoNBpnX65k84Y8clk',
  //   '5rAZLlhh9pQuV6weF2RxGg',
  //   '4kBS1ciuLadz340qeXQGl6',
  //   '3DXocnIyt236ymZAHWYOs6',
  //   'oPv07XcusmLM1UsHniMZy'
  // ]
```
<br />

### `uuidv4` Method  

Creates an API key/access token using random UUID Version 4 generation.  

| Name     | Default Value |  Description                                                      |
| ---------| ------------- | ----------------------------------------------------------------- | 
| `method` | `uuidv4`      | To use the `uuidv4` generation method                             |
| `dashes` | `true`        | Add dashes (`-`) to the API key or not                            |
| `prefix` | `undefined`   | A string prefix for the API key, followed by a period (`.`)       |  
| `batch`  | `undefined`   | The number of API keys to generate                                |
<br />

Examples:  

```javascript
import generateApiKey from 'generate-api-key';

// Provide the generation method.
generateApiKey({ method: 'uuidv4' }); // ⇨ 'c40c974f-307e-490e-8d4e-0c8f31f21df3'

// Create an API key without the dashes.
generateApiKey({ method: 'uuidv4', dashes: false }); // ⇨ 'be1748535cad474191b34a4aa4299a9d'

// Create an API key with a prefix.
generateApiKey({ method: 'uuidv4', prefix: 'test_app' }); // ⇨ 'test_app.1dfdf2c1-7365-4625-b7d9-d9db5210f18d'

// Create a batch (certain amount) of API keys.
generateApiKey({ method: 'uuidv4', batch: 5 }); // ⇨ 
  // [
  //   'd822611a-0600-4edf-9697-21ba5c79e6d7',
  //   '596ac0ae-c4a0-4803-b796-8f239c8431ba',
  //   '0f3819f3-b417-4c4c-b674-853473800265',
  //   '882332bf-4215-41ce-a573-a3c8f5d47a24',
  //   'f269c1f6-77d5-464d-8229-769c0b3a21f7'
  // ]
```
<br />

### `uuidv5` Method  

Creates an API key/access token using random UUID Version 5 generation.  

| Name         | Default Value |  Description                                                      |
| -------------| ------------- | ----------------------------------------------------------------- | 
| `method`     | `uuidv5`      | To use the `uuidv5` generation method                             |
| `name`       | `undefined`   | A unique name to use for the generation                           |
| `namespace`  | `undefined`   | The UUID to use for the generation (ignored if `batch` is given)  |
| `dashes`     | `true`        | Add dashes (`-`) to the API key or not                            |
| `prefix`     | `undefined`   | A string prefix for the API key, followed by a period (`.`)       |  
| `batch`      | `undefined`   | The number of API keys to generate                                |
<br />

Examples:  

```javascript
import generateApiKey from 'generate-api-key';

// Provide the generation method with the name and namespace.
generateApiKey({
  method: 'uuidv5',
  name: 'production app',
  namespace: '1dfdf2c1-7365-4625-b7d9-d9db5210f18d'
}); // ⇨ 'd683c168-377f-528d-8d9a-b7f1551ecb44'

// Create an API key without the dashes.
generateApiKey({
  method: 'uuidv5',
  name: 'production app',
  namespace: '596ac0ae-c4a0-4803-b796-8f239c8431ba',
  dashes: false
}); // ⇨ 'b1bc2cda0c1f5eb594495088a37339b8'

// Create an API key with a prefix.
generateApiKey({
  method: 'uuidv5',
  name: 'production app',
  namespace: '0f3819f3-b417-4c4c-b674-853473800265',
  prefix: 'prod_app'
}); // ⇨ 'prod_app.3f7e5d98-3aa9-5dcb-82e3-10d9a2fc412a'

/* 
 * Create a batch (certain amount) of API keys. When creating a
 * batch of 'uuidv5' API keys, a namespace is generated for each
 * key. The provided namespace is not used.
 */
generateApiKey({
  method: 'uuidv5',
  name: 'production app',
  batch: 5
}); // ⇨ 
  // [
  //   '1e37088a-4dbf-5126-a255-071095e3a53b',
  //   '23c3cfce-2cf1-5b80-856c-12cf6b5f4e88',
  //   '46ec193e-237e-517b-a02e-7679510215d8',
  //   'cf646907-7bc7-5953-bf76-b7527d70b234',
  //   '5ead0c39-0eac-57e8-831d-250932aeb1e0'
  // ]
```
<br />

## Security  

When generating and storing API keys and access tokens please be mindful of secure 
database storage best practices. The reason API keys or access tokens are stored is to 
confirm the key/token that is provided (ex. HTTP request) is valid and issued by your 
organization or application (the same as a password). Just like a password, an API key 
or access token can provide direct access to data or services that require authentication.

To authenticate an API key or access token, it is not necessary to know the raw 
key/token value, the key/token just needs to validated to be correct. API keys and 
access tokens should not be stored in plain text in your database, they should be 
stored as a hashed value. Consider using database storage concepts such as salting 
or peppering during the hashing process. 

Lastly, if you suspect the API credentials for your organization or application have 
been compromised, please revoke the keys and regenerate new keys.  
<br />

## Change Log

The [CHANGELOG](./CHANGELOG.md) contains descriptions of notable changes.  
<br />

## License

This software is licensed under the [Apache 2 license](./LICENSE).  

[npm-url]: https://www.npmjs.com/package/generate-api-key
[version-image]: https://img.shields.io/github/package-json/v/pauldenver/generate-api-key/main?label=version&style=flat-square
[tests-url]: https://github.com/pauldenver/generate-api-key/actions/workflows/test.yml
[tests-image]: https://github.com/pauldenver/generate-api-key/actions/workflows/test.yml/badge.svg?branch=main
[coverage-url]: https://github.com/pauldenver/generate-api-key/actions/workflows/coverage.yml
[coverage-image]: https://github.com/pauldenver/generate-api-key/actions/workflows/coverage.yml/badge.svg?branch=main
[travis-url]: https://travis-ci.com/pauldenver/generate-api-key
[travis-image]: https://travis-ci.com/pauldenver/generate-api-key.svg?branch=main
[coveralls-url]: https://coveralls.io/github/pauldenver/generate-api-key
[coveralls-image]: https://coveralls.io/repos/github/pauldenver/generate-api-key/badge.svg?branch=main
[codefactor-url]: https://www.codefactor.io/repository/github/pauldenver/generate-api-key/overview/main
[codefactor-image]: https://www.codefactor.io/repository/github/pauldenver/generate-api-key/badge/main

CHANGELOG.md
## Changelog  

All notable changes to this project will be documented in this file.  

## [v1.0.2](https://github.com/pauldenver/generate-api-key/compare/v1.0.1...v1.0.2)

### Bug Fixes

*  Updated how the key length is determined when using the `bytes` method ([#3](https://github.com/pauldenver/generate-api-key/issues/3)).
*  Fixed a typo in the `chance.natural` usage in the `string` method.

## [v1.0.1](https://github.com/pauldenver/generate-api-key/compare/v1.0.0...v1.0.1)

### Features

*  Converted the project from JavaScript to TypeScript (also addresses [#1](https://github.com/pauldenver/generate-api-key/issues/1)).
*  Updated module dependencies to their current versions.

### Minor changes

*  Removed the `codecov` NPM module dependency and its associated NPM script. The module has been deprecated.
*  Added config files for `mocha` and `nyc`.

### Breaking changes

*  Updated the supported Node.js engine to versions 14 or greater.

## [v1.0.0](https://github.com/pauldenver/generate-api-key/commit/730c8b2a99d66fa68fb199161422d990228a1161)

*  Initial `generate-api-key` library release.

test/index.js
const { expect } = require('chai');
const rewire = require('rewire');
const { v4: uuidv4 } = require('uuid');
const generateApiKey = require('../lib');

const apiKeyRewire = rewire('../lib/generate_api_key');
const charPool = apiKeyRewire.__get__('CHARS_POOL');
const base62Pool = apiKeyRewire.__get__('BASE62_POOL');

describe('generateApiKey', () => {
  context(`'bytes' method`, () => {
    it('should create an API key with the default options', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'bytes' });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf.within(16, 32);
    });

    it('should create an API key with a set length', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'bytes', length: 10 });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(10);
    });

    it('should create an API key with an uneven set length', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'bytes', length: 7 });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(7);
    });

    it(`should create an API key with a length within the 'min' and 'max' range`, () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'bytes', min: 5, max: 15 });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf.within(5, 15);
    });

    it(`should create an API key with an uneven 'min' and 'max' range`, () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'bytes', min: 15, max: 15 });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(15);
    });
  });

  context(`'string' method`, () => {
    it('should create an API key with the default options', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'string' });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf.within(16, 32);
      [ ...apiKey ].forEach((c) => expect(charPool).to.include(c));
    });

    it('should create an API key with a set length', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'string', length: 15 });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(15);
      [ ...apiKey ].forEach((c) => expect(charPool).to.include(c));
    });

    it(`should create an API key with a length within the 'min' and 'max' range`, () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'string', min: 10, max: 25 });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf.within(10, 25);
      [ ...apiKey ].forEach((c) => expect(charPool).to.include(c));
    });
  });

  context(`'base32' method`, () => {
    it('should create an API key with the default options', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'base32' });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(31);
      expect(apiKey).to.match(/[A-Z0-9]{7}-[A-Z0-9]{7}-[A-Z0-9]{7}-[A-Z0-9]{7}/);
    });

    it('should create an API key without dashes', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'base32', dashes: false });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(28);
      expect(apiKey).to.match(/[A-Z0-9]{28}/);
    });
  });

  context(`'base62' method`, () => {
    it('should create an API key', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'base62' });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf.at.most(22);
      [ ...apiKey ].forEach((c) => expect(base62Pool).to.include(c));
    });
  });

  context(`'uuidv4' method`, () => {
    it('should create an API key with the default options', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'uuidv4' });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(36);
      expect(apiKey).to.match(
        /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/
      );
    });

    it('should create an API key without dashes', () => {
      // Create the API key.
      const apiKey = generateApiKey({ method: 'uuidv4', dashes: false });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(32);
      expect(apiKey).to.match(/[0-9a-fA-F]{32}/);
    });
  });

  context(`'uuidv5' method`, () => {
    it('should create an API key with the default options', () => {
      // Create the API key.
      const apiKey = generateApiKey({
        method: 'uuidv5',
        name: 'Super secret',
        namespace: uuidv4(),
      });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(36);
      expect(apiKey).to.match(
        /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/
      );
    });

    it('should create API keys with a created namespace in batch mode', () => {
      // Create the API keys.
      const apiKeys = generateApiKey({
        method: 'uuidv5',
        name: 'Super secret',
        batch: 2,
      });

      expect(apiKeys).to.be.an('array');
      expect(apiKeys).to.have.lengthOf(2);

      apiKeys.forEach((key) => {
        expect(key).to.be.a('string');
        expect(key).to.have.lengthOf(36);
        expect(key).to.match(
          /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/
        );
      });
    });

    it('should create an API key without dashes', () => {
      // Create the API key.
      const apiKey = generateApiKey({
        method: 'uuidv5',
        name: 'Super secret',
        namespace: uuidv4(),
        dashes: false,
      });
      
      expect(apiKey).to.be.a('string');
      expect(apiKey).to.have.lengthOf(32);
      expect(apiKey).to.match(/[0-9a-fA-F]{32}/);
    });

    it(`should throw an error for an invalid 'name' value`, () => {
      expect(() => generateApiKey({
        method: 'uuidv5',
        namespace: uuidv4(),
      })).to.throw(TypeError, `The required 'name' option must be a string.`);

      expect(() => generateApiKey({
        method: 'uuidv5',
        name: 102,
        namespace: uuidv4(),
      })).to.throw(TypeError, `The required 'name' option must be a string.`);
    });

    it(`should throw an error for an invalid 'namespace' value`, () => {
      expect(() => generateApiKey({
        method: 'uuidv5',
        name: 'Very special',
      })).to.throw(TypeError, `The required 'namespace' option must be a string.`);

      expect(() => generateApiKey({
        method: 'uuidv5',
        name: 'Very special',
        namespace: {},
      })).to.throw(TypeError, `The required 'namespace' option must be a string.`);
    });
  });

  it('should create an API key with a prefix', () => {
    // Create the API key.
    const apiKey = generateApiKey({ method: 'string', prefix: 'app' });
    
    expect(apiKey).to.be.a('string');
    expect(apiKey.startsWith('app.')).to.be.true;
    expect(apiKey.replace(/^app\./, '')).to.have.lengthOf.within(16, 32);
    [ ...apiKey.replace(/^app\./, '') ].forEach((c) => expect(charPool).to.include(c));
  });

  it(`should create an API key using the 'string' method when not provided`, () => {
    // Create the API key.
    const apiKey = generateApiKey();
    
    expect(apiKey).to.be.a('string');
    expect(apiKey).to.have.lengthOf.within(16, 32);
    [ ...apiKey ].forEach((c) => expect(charPool).to.include(c));
  });

  it(`should throw an error for an invalid 'method' value`, () => {
    expect(() => generateApiKey({
      method: 99,
    })).to.throw(Error, 'Received an unknown API key generation method.');

    expect(() => generateApiKey({
      method: 'other',
      batch: 8,
    })).to.throw(Error, 'Received an unknown API key generation method.');
  });

  it('should generate a batch of API keys', () => {
    // Create the API keys.
    const apiKeys = generateApiKey({ method: 'string', batch: 10 });
    
    expect(apiKeys).to.be.an('array');
    expect(apiKeys).to.have.lengthOf(10);

    apiKeys.forEach((key) => {
      expect(key).to.be.a('string');
      expect(key).to.have.lengthOf.within(16, 32);
      [ ...key ].forEach((c) => expect(charPool).to.include(c));
    });
  });

  it('should generate a batch of API keys with a prefix', () => {
    // Create the API keys.
    const apiKeys = generateApiKey({ method: 'string', batch: 5, prefix: 'test' });
    
    expect(apiKeys).to.be.an('array');
    expect(apiKeys).to.have.lengthOf(5);

    apiKeys.forEach((key) => {
      expect(key).to.be.a('string');
      expect(key.startsWith('test.')).to.be.true;
      expect(key.replace(/^test\./, '')).to.have.lengthOf.within(16, 32);
      [ ...key.replace(/^test\./, '') ].forEach((c) => expect(charPool).to.include(c));
    });
  });

  it(`should throw an error for an invalid 'batch' value`, () => {
    expect(() => generateApiKey({
      method: 'string',
      batch: 4.25,
    })).to.throw(TypeError, `The 'batch' option must be a natural number > 0.`);

    expect(() => generateApiKey({
      method: 'string',
      batch: -5,
    })).to.throw(TypeError, `The 'batch' option must be a natural number > 0.`);

    expect(() => generateApiKey({
      method: 'string',
      batch: [],
    })).to.throw(TypeError, `The 'batch' option must be a natural number > 0.`);
  });
});

.eslintrc.js
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint'
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  rules: { 
    'no-console': 'error',
    'max-len': [ 'error', 120, 2, {
      ignoreUrls: true,
      ignoreComments: false,
      ignoreRegExpLiterals: true,
      ignoreStrings: true,
      ignoreTemplateLiterals: true,
    }],
    'semi': [ 'error', 'always' ],
    '@typescript-eslint/no-explicit-any': 'off',
  },
  overrides: [
    {
      files: [ 'test/generate_api_key.spec.ts' ],
      rules: {
        '@typescript-eslint/no-var-requires': 'off',
        '@typescript-eslint/ban-ts-comment': [
          'error',
          {
            'ts-ignore': false,
          },
        ],
      },
    }
  ]
};

lib/generate_api_key.js
const Chance = require('chance');
const { v4: uuidv4, v5: uuidv5 } = require('uuid');
const { randomBytes } = require('crypto');
const { base32 } = require('rfc4648');
const baseX = require('base-x');

// Character pools.
const LOWER_POOL = 'abcdefghijklmnopqrstuvwxyz';
const UPPER_POOL = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const NUM_POOL = '0123456789';
const SPECIAL_POOL = '-._~+/';

// Create the combined character pool.
const CHARS_POOL = `${LOWER_POOL}${UPPER_POOL}${NUM_POOL}${SPECIAL_POOL}`;
// Create the Base62 character pool.
const BASE62_POOL = `${NUM_POOL}${LOWER_POOL}${UPPER_POOL}`;

// Helper function for checking for a string value.
const isString = (value) => {
  return (typeof value === 'string' || value instanceof String);
};

// Helper function for checking for a natural number.
const isNaturalNum = (value) => {
  // Check the type.
  if (typeof value !== 'number') {
    return false;
  } else {
    if (value >= 0.0) {
      return (Math.floor(value) === value) && value !== Infinity;
    }

    return false;
  }
};

// Helper function for getting the options.
const getOptions = (options, defaults) => {
  Object.keys(defaults).forEach((key) => {
    // Assign the value if not already set.
    if (options[key] === undefined) {
      options[key] = defaults[key];
    }
  });

  return options;
};

/**
 * Creates an API key using random bytes.
 *
 * @param {Object} options API key options.
 * @returns {String} The API key.
 */
function _getCryptoApiKey(options) {
  let totalBytes = 16;
  let apiKey;

  // Get the options.
  options = getOptions(options, { min: 16, max: 32 });

  // Get a 'Chance' instance.
  const chance = new Chance();

  if (options.length) {
    totalBytes = Math.ceil(options.length / 2);
  } else {
    // Get a random number.
    const numVal = chance.natural({ min: options.min, max: options.max });

    // Set the total bytes.
    totalBytes = Math.ceil(numVal / 2);
  }

  // Generate the API key.
  apiKey = randomBytes(totalBytes).toString('hex');

  // Check the key length.
  if (options.length && (apiKey.length > options.length)) {
    const endIndex = apiKey.length - (apiKey.length - options.length);
    apiKey = apiKey.slice(0, endIndex);
  } else if (apiKey.length > options.max) {
    const endIndex = apiKey.length - (apiKey.length - options.max);
    apiKey = apiKey.slice(0, endIndex);
  }

  return apiKey;
}

/**
 * Creates an API key using random string generation.
 *
 * @param {Object} options API key options.
 * @returns {String} The API key.
 */
function _getRandomStringApiKey(options) {
  // Get the options.
  options = getOptions(options, { min: 16, max: 32, pool: CHARS_POOL });

  // Get a 'Chance' instance.
  const chance = new Chance();

  // Generate the string.
  return chance.string({
    length: options.length || undefined,
    min: options.min,
    max: options.max,
    pool: options.pool,
  });
}

/**
 * Creates an API key using Base32 Crockford encoding.
 *
 * @param {Object} options API key options.
 * @returns {String} The API key.
 */
function _getBase32CrockfordApiKey(options) {
  // Get the options.
  options = getOptions(options, { dashes: true });

  // Create the uuid options.
  const v4options = {
    random: randomBytes(16),
  };

  // Create a new UUID.
  const uuid = uuidv4(v4options);
  // Split at the dashes.
  const uuidParts = uuid.split('-');

  // Convert the UUID into 4 equally separate parts.
  const partsArr = [
    uuidParts[0],
    `${uuidParts[1]}${uuidParts[2]}`,
    `${uuidParts[3]}${uuidParts[4].substr(0, 4)}`,
    uuidParts[4].substr(4),
  ];

  let apiKey = '';

  // Iterate through each part.
  partsArr.forEach((value) => {
    // Get every two characters.
    const valueArr = value.match(/.{1,2}/g);

    // Convert each value into a number.
    const numArr = valueArr.map((item) => Number(`0x${item}`));

    // Create the string.
    const base32Str = base32.stringify(numArr, { pad: false });

    // Check if we should add dashes
    if (options.dashes) {
      if (apiKey.length === 0) {
        apiKey += base32Str;
      } else {
        apiKey += `-${base32Str}`;
      }
    } else {
      apiKey += base32Str;
    }
  });

  return apiKey;
}

/**
 * Creates an API key using Base62 encoding.
 *
 * @returns {String} The API key.
 */
function _getBase62ApiKey() {
  // Create the uuid options.
  const v4options = {
    random: randomBytes(16),
  };

  // Set the encoding alphabet for Base62.
  const base62 = baseX(BASE62_POOL);

  // Create a new UUID.
  const uuid = uuidv4(v4options);
  // Create the UUID buffer.
  const uuidBuffer = Buffer.from(uuid.replace(/-/g, ''), 'hex');

  // Generate the API key.
  const apiKey = base62.encode(uuidBuffer);

  return apiKey;
}

/**
 * Creates an API key using random UUID Version 4 generation.
 *
 * @param {Object} options API key options.
 * @returns {String} The API key.
 */
function _getUuidV4ApiKey(options) {
  // Get the options.
  options = getOptions(options, { dashes: true });

  // Create the uuid options.
  const v4options = {
    random: randomBytes(16),
  };

  // Generate the API key.
  const apiKey = uuidv4(v4options);

  // Check if we should remove dashes.
  if (!options.dashes) {
    return apiKey.replace(/-/g, '');
  }

  return apiKey;
}

/**
 * Creates an API key using random UUID Version 5 generation.
 *
 * @param {Object} options API key options.
 * @returns {String} The API key.
 */
function _getUuidV5ApiKey(options) {
  // Get the options.
  options = getOptions(options, { dashes: true });

  let namespace;

  if (!isString(options.name)) {
    throw new TypeError(`The required 'name' option must be a string.`);
  }

  // Check of batch processing. If so, create a namespace UUID.
  if (options.batch) {
    // Create the uuid options.
    const v4options = {
      random: randomBytes(16),
    };

    // Generate a namespace UUID.
    namespace = uuidv4(v4options);
  } else {
    if (!isString(options.namespace)) {
      throw new TypeError(`The required 'namespace' option must be a string.`);
    }

    namespace = options.namespace;
  }

  // Generate the API key.
  const apiKey = uuidv5(options.name, namespace);

  // Check if we should remove dashes.
  if (!options.dashes) {
    return apiKey.replace(/-/g, '');
  }

  return apiKey;
}

/**
 * Generates a simple API key based on the provided
 * generation method.
 *
 * @param {Object} options The API key generation options.
 * @returns {String|Array<String>} The API key or array of API keys.
 */
function generateApiKey(options = {}) {
  // Get the generation method.
  const genMethod = options.method || 'string';

  // Check for batch generation.
  if (options.batch) {
    if (!isNaturalNum(options.batch) || options.batch === 0) {
      throw new TypeError(`The 'batch' option must be a natural number > 0.`);
    }
  }

  // An object literal for creating the API key.
  const getApiKey = (method) => ({
    bytes: () => _getCryptoApiKey(options),
    string: () => _getRandomStringApiKey(options),
    base32: () => _getBase32CrockfordApiKey(options),
    base62: () => _getBase62ApiKey(),
    uuidv4: () => _getUuidV4ApiKey(options),
    uuidv5: () => _getUuidV5ApiKey(options),
    default: () => {
      throw new Error('Received an unknown API key generation method.');
    },
  })[method];

  // Check for batch generation.
  if (options.batch) {
    const apiKeys = [];

    for (let i = 0; i < options.batch; i++) {
      // Generate the API key.
      let apiKey = (getApiKey(genMethod) || getApiKey('default'))();

      // Add a prefix if necessary.
      if (options.prefix) {
        apiKey = `${options.prefix}.${apiKey}`;
      }

      // Add the key to the array.
      apiKeys.push(apiKey);
    }

    return apiKeys;
  } else {
    // Generate the API key.
    const apiKey = (getApiKey(genMethod) || getApiKey('default'))();

    // Add a prefix if necessary.
    if (options.prefix) {
      return `${options.prefix}.${apiKey}`;
    }

    return apiKey;
  }
}

module.exports = generateApiKey;
